client/components/marked-lines/index.js

```diff
@@ -0,0 +1,86 @@
+/** @format */
+/**
+ * External dependencies
+ */
+import React from 'react';
+import classNames from 'classnames';
+import { map } from 'lodash';
+
+/**
+ * Surrounds a text string in a <mark>
+ * Just a small helper function
+ *
+ * @example
+ * mark( 'be kind' ) =>
+ *   <mark key="be kind" className="marked-lines__mark">be kind</mark>
+ *
+ * @param {string} text the string to mark
+ * @return {Element} React <mark> Element
+ */
+const mark = text => (
+	<mark key={ text } className="marked-lines__mark">
+		{ text }
+	</mark>
+);
+
+/**
+ * Translates marked-file context input
+ * into React component output
+ *
+ * @example
+ * const marks = [ [ 2, 4 ], [ 5, 9 ] ]
+ * const content = '->^^-_____<--'
+ * markup( marks, content ) === [ '->', <mark>{ '^^' }</mark>, '-', <mark>{ '_____' }</mark>, '<--' ]
+ *
+ * @param {Array<Array<number>>} marks spanning indices of text to mark, values in UCS-2 code units
+ * @param {string} content the plaintext content to mark
+ * @return {Array|string} list of output text nodes and mark elements or plain string output
+ */
+const markup = ( marks, content ) => {
+	const [ finalOutput, finalLast ] = marks.reduce(
+		( [ output, lastIndex ], [ markStart, markEnd ] ) => {
+			// slice of input text specified by current mark ranges
+			const slice = content.slice( markStart, markEnd );
+
+			// if we have text before the first index then prepend it as well
+			const next =
+				markStart > lastIndex
+					? [ content.slice( lastIndex, markStart ), mark( slice ) ]
+					: [ mark( slice ) ];
+
+			return [ [ ...output, ...next ], markEnd ];
+		},
+		[ [], 0 ]
+	);
+
+	// we may also have text after the last mark
+	return finalLast < content.length ? [ ...finalOutput, content.slice( finalLast ) ] : finalOutput;
+};
+
+const MarkedLines = ( { context } ) => {
+	const { marks, ...lines } = context;
+
+	return (
+		<div className="marked-lines">
+			{ map( lines, ( content, lineNumber ) => {
+				const hasMarks = marks.hasOwnProperty( lineNumber );
+
+				return (
```

Comment:
Similar to my comment about the color that should be moved to a class, there should be some prop that can be passed to the range that allows to specify the highlighting.

Reply:
for completeness I'll answer here too - if there's a need I've overlooked I think it's good what you are suggesting but I'm not sure we have identified any such need yet. I'm all 👂s though for things I may have overlooked